instrumentation/opentelemetry-instrumentation-openai-v2/src/opentelemetry/instrumentation/openai_v2/utils.py

```diff
@@ -43,69 +48,138 @@ def wrapper(*args, **kwargs):
-            }
+        func = get_property_value(tool_call, "function")
+        if func:
+            tool_call_dict["function"] = {}
+
+            name = get_property_value(func, "name")
+            if name:
+                tool_call_dict["function"]["name"] = name
+
+            arguments = get_property_value(func, "arguments")
+            if arguments:
+                tool_call_dict["function"]["arguments"] = arguments.replace(
+                    "\n", ""
+                )
+
         calls.append(tool_call_dict)
     return calls
 
 
-def set_event_prompt(span, prompt):
-    span.add_event(
-        name="gen_ai.content.prompt",
-        attributes={
-            GenAIAttributes.GEN_AI_PROMPT: prompt,
-        },
+def set_server_address_and_port(client_instance, attributes):
+    base_client = getattr(client_instance, "_client", None)
+    base_url = getattr(base_client, "base_url", None)
+    if not base_url:
+        return
+
+    port = -1
+    if isinstance(base_url, URL):
+        attributes[ServerAttributes.SERVER_ADDRESS] = base_url.host
+        port = base_url.port
+    elif isinstance(base_url, str):
+        try:
+            url = urlparse(base_url)
+            attributes[ServerAttributes.SERVER_ADDRESS] = url.hostname
+            port = url.port
+        except Exception:
+            pass
+
+    if port and port != 443 and port > 0:
+        attributes[ServerAttributes.SERVER_PORT] = port
+
+
+def get_property_value(obj, property_name):
+    if isinstance(obj, dict):
+        return obj.get(property_name, None)
+
+    return getattr(obj, property_name, None)
+
+
+def message_to_event(message, span_ctx):
+    attributes = {
+        GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value
+    }
+    role = get_property_value(message, "role")
+    content = get_property_value(message, "content")
+
+    body = {}
+    if content:
+        body["content"] = content
+    if role == "assistant":
+        tool_calls = extract_tool_calls(message)
+        if tool_calls:
+            body = {"tool_calls": tool_calls}
+    elif role == "tool":
+        tool_call_id = get_property_value(message, "tool_call_id")
+        if tool_call_id:
+            body["id"] = tool_call_id
+
+    return Event(
+        name=f"gen_ai.{role}.message",
+        attributes=attributes,
+        body=body,
+        trace_id=span_ctx.trace_id,
+        span_id=span_ctx.span_id,
+        trace_flags=span_ctx.trace_flags,
+    )
+
+
+def choice_to_event(choice, span_ctx):
+    attributes = {
+        GenAIAttributes.GEN_AI_SYSTEM: GenAIAttributes.GenAiSystemValues.OPENAI.value
+    }
+
+    body = {
+        "index": choice.index,
+        "finish_reason": choice.finish_reason or "error",
+    }
+
+    if choice.message:
+        message = {
+            "role": choice.message.role
+            if choice.message and choice.message.role
+            else "assistant"
```

Comment:
> Conditionally Required: if available and if not equal to assistant

Does that mean if it IS equal to "assistant", we do not populate this attribute? From the sem conv, it's not obvious to me that "assistant" is the default value if nothing is there.

Reply:
<img width="1187" alt="image" src="https://github.com/user-attachments/assets/c700d208-23eb-4078-84c6-023c655523d4">

https://github.com/open-telemetry/semantic-conventions/blob/v1.28.0/docs/gen-ai/gen-ai-events.md#assistant-event

Reply:
good catch, removed default